### PR TITLE
[remoting] Convert mono_load_remote_field_new_checked to managed code

### PR DESCRIPTION
**Pull request by lambdageek**
Also surface `MonoRemoteClass` struct to managed code.

### CODE REVIEW

**Comment by lambdageek** [comment]
Attn @kumpera @vargaz @marek-safar 
Please let me know what I'm doing wrong.  I'll be doing a bunch of these, so better to catch bad manners sooner.

File: mcs/class/corlib/Mono/RuntimeMarshaling.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Copyright 2016 Dot net foundation.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Mono {
+	internal sealed class RuntimeMarshaling {
```
Comment (marek-safar) [inline mcs/class/corlib/Mono/RuntimeMarshaling.cs]: Should be static

File: mcs/class/corlib/Mono/RuntimeMarshaling.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		private unsafe extern static RuntimeTypeHandle mono_class_get_type (RuntimeStructs.MonoClass* klass);
+
+		internal unsafe static RuntimeTypeHandle RuntimeTypeHandleFromClass (RuntimeStructs.MonoClass* klass)
```
Comment (marek-safar) [inline mcs/class/corlib/Mono/RuntimeMarshaling.cs]: Do we need extra wrapper, cannot it be just properly named icall?

File: mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs [hunk 3]
```diff
@@ -109,19 +109,30 @@ public MonoMethodMessage (MethodBase method, object [] out_args)
...
-			InitMessage ((MonoMethod)minfo, null);
+			InitMessage ((MonoMethod)minfo, out_args);
 
 			int len = in_args.Length;
 			for (int i = 0; i < len; i++) {
 				args [i] = in_args [i];
 			}
 		}
+
+		private static MethodInfo getMethodInfo (Type type, string method_name)
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs]: Use C# naming conventions

File: mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs [hunk 4]
```diff
@@ -109,19 +109,30 @@ public MonoMethodMessage (MethodBase method, object [] out_args)
...
+		{
+			MethodInfo minfo = type.GetMethod(method_name);
+			if (minfo == null)
+				throw new ExecutionEngineException (String.Format("Could not find '{0}' in {1}", method_name, type));
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs]: Throw ArgumentException instead

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 5]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
 	[StructLayout (LayoutKind.Sequential)]
 	internal class TransparentProxy {
 		public RealProxy _rp;
-		IntPtr _class;
+		unsafe Mono.RuntimeStructs.RemoteClass* _class;
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: Cannot this be of RuntimeTypeHandle?

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 6]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
...
+
+		bool IsContextBoundObject {
+			get { return proxy_type ().IsContextful; }
+		}
+
+		Context TargetContext {
+			get { return _rp._targetContext; }
+		}
+
+		internal object load_remote_field_new (RuntimeTypeHandle type_handle, RuntimeFieldHandle field_handle) {
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: Use C# naming convention

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 7]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
...
+			if (inCurrCtx) {
+				object o = _rp._server;
+				return field.GetValue(o);
+			}
+
+			object[] in_args = new object[] { Type.GetTypeFromHandle(type_handle).FullName,
+							  field.Name };
+			MethodInfo minfo = typeof(object).GetMethod("FieldGetter", BindingFlags.NonPublic | BindingFlags.Instance);
+			if (minfo == null)
+				throw new ExecutionEngineException ("Couldn't find FieldGetter in object when making a method message");
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: Use System.MissingMethodException instead

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs (hunk 5 above)
Comment (lambdageek) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: A `RemoteClass` is not a `MonoType` (aka `RuntimeTypeHandle`), actually it's not even a `MonoClass`, it's a separate ad-hoc structure.  I can wrap it up in a value type like `RuntimeTypeHandle` wraps up `MonoType*`, though.

File: mcs/class/corlib/Mono/RuntimeStructs.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		[StructLayout(LayoutKind.Sequential)]
+		internal unsafe struct RemoteClass {
+			internal IntPtr default_vtable;
+			internal IntPtr xdomain_vtable;
+			internal MonoClass* proxy_class;
+			internal IntPtr proxy_class_name;
+			internal UInt32 interface_count;
+			// FIXME: How to represent variable-length array struct member?
+			// MonoClass* interfaces [];
+		}
```
Comment (migueldeicaza) [inline mcs/class/corlib/Mono/RuntimeStructs.cs]: Isn't this a major blocker to get this landed?

File: mcs/class/corlib/Mono/RuntimeStructs.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+namespace Mono {
+	internal class RuntimeStructs {
+		// class-internals.h MonoRemoteClass
+		[StructLayout(LayoutKind.Sequential)]
+		internal unsafe struct RemoteClass {
+			internal IntPtr default_vtable;
+			internal IntPtr xdomain_vtable;
+			internal MonoClass* proxy_class;
+			internal IntPtr proxy_class_name;
+			internal UInt32 interface_count;
```
Comment (kumpera) [inline mcs/class/corlib/Mono/RuntimeStructs.cs]: use uint.

File: mcs/class/corlib/Mono/RuntimeStructs.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+			// FIXME: How to represent variable-length array struct member?
```
Comment (kumpera) [inline mcs/class/corlib/Mono/RuntimeStructs.cs]: How about a private IntPtr for the first address of the variable area plus a bunch of properties/functions to fetch all elements?

File: mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs (hunk 3 above)
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Messaging/MonoMethodMessage.cs]: Meaning it should be: static MethodInfo GetMethodInfo (Type type, string methodName)

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs (hunk 5 above)
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: What about we model this as follow:

Mono.RuntimeStructs.* < straight mapping of native structs
Mono.RuntimeHandles.* < wrapper Handle variants that contain any methods we might need.

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 11]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
 	[StructLayout (LayoutKind.Sequential)]
 	internal class TransparentProxy {
 		public RealProxy _rp;
-		IntPtr _class;
+		unsafe Mono.RuntimeStructs.RemoteClass* _class;
 		bool _custom_type_info;
+
+		unsafe internal RuntimeType proxy_type() {
```
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: Calling convention

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 12]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
...
+		}
+
+		internal object load_remote_field_new (RuntimeTypeHandle type_handle, RuntimeFieldHandle field_handle) {
+
+			FieldInfo field = FieldInfo.GetFieldFromHandle (field_handle);
+			bool inCurrCtx =
+				IsContextBoundObject && TargetContext == Thread.CurrentContext;
+			if (inCurrCtx) {
+				object o = _rp._server;
+				return field.GetValue(o);
```
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: wow, code goes tons simpler when you have reflection available.

File: mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs [hunk 13]
```diff
@@ -49,8 +49,45 @@ namespace System.Runtime.Remoting.Proxies
...
+			}
+
+			object[] in_args = new object[] { Type.GetTypeFromHandle(type_handle).FullName,
+							  field.Name };
+			MethodInfo minfo = typeof(object).GetMethod("FieldGetter", BindingFlags.NonPublic | BindingFlags.Instance);
+			if (minfo == null)
+				throw new ExecutionEngineException ("Couldn't find FieldGetter in object when making a method message");
+			MonoMethodMessage msg = new MonoMethodMessage (minfo, in_args, null);
```
Comment (kumpera) [inline mcs/class/corlib/System.Runtime.Remoting.Proxies/RealProxy.cs]: Why aren't you creating a size one object[] array and passing to MonoMethodMessage?

**Comment by lambdageek** [comment]
Revised according to @marek-safar and @kumpera 's comments.

**Comment by lambdageek** [comment]
While I'm at it, looks like I can entirely get rid of the `MONO_WRAPPER_LDFLD_REMOTE` wrapper type

**Comment by kumpera** [comment]
When we change any of the aot constants (numeric value of wrappers in this case), we must bump the aot format version.

Besides that minor point LGTM.

Great catch on removing the wrapper completely.

**Comment by lambdageek** [comment]
@monojenkins merge

**Comment by lambdageek** [comment]
Test failures look unrelated.

**Pull request merged**